### PR TITLE
feat: querier test system, ground work

### PR DESCRIPTION
**Pull request by crepererum**
See #3985 for the motivation.

This introduces a cache system for the querier which can later be
extended to support the remaining features listed in #3985 (e.g.
metrics, LRU/TTL).

All current caches are wired up to go throw the new cache system. Once
we move away from (ab)using `db`, the set of caches will be different
but the system will remain.

### CODE REVIEW

File: querier/src/cache_system/broadcast.rs [hunk 1]
```diff
@@ -0,0 +1,144 @@
+use observability_deps::tracing::warn;
```
Comment (crepererum) [inline querier/src/cache_system/broadcast.rs:1]: This file is copied from `rskafka` where we had a similar application.

File: querier/src/cache_system/driver.rs [hunk 2]
```diff
@@ -0,0 +1,430 @@
+use std::{collections::HashMap, future::Future, hash::Hash, sync::Arc};
```
Comment (crepererum) [inline querier/src/cache_system/driver.rs:1]: This is the actual cache system.

File: querier/src/cache_system/driver.rs [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+        }
+    }
+}
+
+/// Inner cache state that is usually guarded by a lock.
+///
+/// The state parts must be updated in a consistent manner, i.e. while using the same lock guard.
+#[derive(Debug)]
+struct CacheState<K, V> {
+    cached_entries: HashMap<K, V>,
```
Comment (crepererum) [inline querier/src/cache_system/driver.rs:140]: Currently we store everything in a "keep all" hash map. That's obviously not the desired end state. I'll introduce an interface for that bit (that also takes care of pruning entries) later, but I didn't want to blow up this PR any further.

File: querier/src/cache/mod.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
+use backoff::BackoffConfig;
```
Comment (crepererum) [inline querier/src/cache/mod.rs:1]: Everything in `querier::cache` is actually just moved around code. The caches existed beforehand but in a giant struct and hand-managed hashmaps. They are not split apart. The tests are the same (just moved to the corresponding modules).

File: querier/src/cache_system/driver.rs (hunk 2 above)
Comment (crepererum) [reply querier/src/cache_system/driver.rs:1]: @tustvold I could use your review here.

**Comment by crepererum** [comment]
FYI: @tustvold proposed to just use a shared oneshot channel instead of `BroadcastOnce`. This would also avoid the usage of the weird panic guard. I'll implement that tomorrow. This shouldn't affect the interface or the tests though.

**Comment by tustvold** [comment]
Looks good to me, nice work. As you've articulated, whilst I like the idea of reusing BroadcastOnce, I think the fact we don't need peek() functionality and already have external state storage means we can do something simpler

File: querier/src/cache_system/broadcast.rs (hunk 1 above)
Comment (crepererum) [reply querier/src/cache_system/broadcast.rs]: this file is gone now

File: querier/src/cache_system/driver.rs [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+            state,
+            k,
+            did_not_panic: false,
+        }
+    }
+
+    /// Flag that the loader did not panic.
+    ///
+    /// Dropping this guard after calling this function will NOT broadcast anything.
+    fn did_not_panic(&mut self) {
```
[the remark this answers is not in the document]
Comment (crepererum) [reply querier/src/cache_system/driver.rs]: this code is gone now

File: querier/src/cache_system/driver.rs [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+
+impl<K, V> Cache<K, V>
+where
+    K: Clone + Eq + Hash + std::fmt::Debug + Send + 'static,
+    V: Clone + std::fmt::Debug + Send + 'static,
+{
+    /// Create new, empty cache with given loader function.
+    pub fn new<T, F>(loader: T) -> Self
+    where
+        T: Fn(K) -> F + Send + Sync + 'static,
```
[the remark this answers is not in the document]
Comment (crepererum) [reply querier/src/cache_system/driver.rs]: done

File: querier/src/cache_system/driver.rs [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+
+        tokio::time::sleep(Duration::from_millis(10)).await;
+
+        assert_eq!(Arc::strong_count(&loader), 1);
+    }
+
+    fn setup() -> (Arc<Cache<u8, String>>, Arc<Loader>) {
+        let loader = Arc::new(Loader::default());
+        let loader_captured = Arc::clone(&loader);
+        let cache = Arc::new(Cache::new(move |key| {
```
[the remark this answers is not in the document]
Comment (crepererum) [reply querier/src/cache_system/driver.rs]: done

**Comment by crepererum** [comment]
@tustvold I've removed `BroadcastOnce`, which resulted in way simpler and far less code. I've also introduced a `Loader` trait. So PTAL :slightly_smiling_face:

**Review by tustvold — APPROVED** [review]

File: querier/src/cache_system/driver.rs [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+                    // If we panic here then `tx` will be dropped and the receivers will be notified.
+                    let v = loader.load(k_for_loader).await;
+
+                    // broadcast result
+                    // It's OK if the receiver side is gone. This might happen during shutdown
+                    tx.send(v.clone()).ok();
+
+                    // remove "running" state and store result
+                    let mut state = state_captured.lock();
+                    drop(
```
[the remark this answers is not in the document]
Comment (crepererum) [reply querier/src/cache_system/driver.rs:90]: I'll add a comment. In short: to convince `rustc` that we don't need the shared future that is returned by `remove`.

**Pull request merged**